numba/ir_utils.py

```diff
@@ -1965,6 +1966,240 @@ def warn_deprecated(func_ir, typemap):
+                                                block, i, pyfunc,
+                                                work_list=work_list)
+                            return True
+        return False
+
+
+class InlineOverloads(object):
+
+    _DEBUG = False
+
+    def __init__(self, func_ir, tyctx, cgctx, type_annotation):
+        self.func_ir = func_ir
+        self.tyctx = tyctx
+        self.cgctx = cgctx
+        self.type_annotation = type_annotation
+        self.typemap = type_annotation.typemap
+        self.calltypes = type_annotation.calltypes
+
+    def run(self):
+        """Run inlining of overloads
+        """
+        if config.DEBUG or self._DEBUG:
+            print('before overload inline'.center(80, '-'))
+            print(self.func_ir.dump())
+            print(''.center(80, '-'))
+        modified = False
+        work_list = list(self.func_ir.blocks.items())
+        while work_list:
+            label, block = work_list.pop()
+            for i, instr in enumerate(block.body):
+                if isinstance(instr, ir.Assign):
+                    expr = instr.value
+                    if isinstance(expr, ir.Expr) and expr.op == 'call':
+                        if guard(self._do_work, work_list, block, i, expr):
+                            modified = True
+                            break # because block structure changed
+
+        if modified:
+            dead_code_elimination(self.func_ir, typemap=self.typemap)
+            # clean up unconditional branches that appear due to inlined
+            # functions introducing blocks
+            self.func_ir.blocks = simplify_CFG(self.func_ir.blocks)
+
+        if config.DEBUG or self._DEBUG:
+            print('after overload inline'.center(80, '-'))
+            print(self.func_ir.dump())
+            print(''.center(80, '-'))
+
+    def _do_work(self, work_list, block, i, expr):
+        from numba.inline_closurecall import inline_closure_call
+
+        # try and get a definition for the call, this isn't always possible as
+        # it might be a eval(str)/part generated awaiting update etc. (parfors)
+        to_inline = None
+        try:
+            to_inline = self.func_ir.get_definition(expr.func)
+        except Exception as e:
+            return False
+
+        # do not handle closure inlining here, another pass deals with that.
+        if getattr(to_inline, 'op', False) == 'make_function':
+            return False
+
+        func_ty = self.typemap[expr.func.name]
+        if not hasattr(func_ty, 'get_call_type'):
+            return False
+
+
+        templates = getattr(func_ty, 'templates', None)
+        if templates is None:
+            return False
+
+        sig = self.calltypes[expr]
+
+        impl = None
+        for template in templates:
+            inline_type = getattr(template, '_inline', 'never')
+            if inline_type != 'never':
+                try:
+                    impl = template._overload_func(*sig.args)
```

Comment:
I think this needs `assert impl is not None`.

Reply:
good catch